pkg/endpoint/bpf.go

```diff
@@ -903,48 +904,138 @@ func (e *Endpoint) GetBPFValue() (*lxcmap.EndpointInfo, error) {
+		e.getLogger().WithError(err).WithFields(logrus.Fields{
+			logfields.BPFMapKey: policymapKey,
+			logfields.Port:      entry.ProxyPort,
+		}).Error("Failed to add PolicyMap key")
+		return err
+	}
+
+	// Operation was successful, add to realized state.
+	e.realizedPolicy.PolicyMapState[keyToAdd] = entry
+
+	// Incremental updates need to update the desired state as well.
+	if incremental && e.desiredPolicy != e.realizedPolicy {
+		e.desiredPolicy.PolicyMapState[keyToAdd] = entry
+	}
+
+	return nil
+}
+
+func (e *Endpoint) proxyID(key policy.Key) string {
+	return policy.ProxyID(e.ID, key.TrafficDirection == trafficdirection.Ingress.Uint8(), u8proto.U8proto(key.Nexthdr).String(), key.DestPort)
+}
+
+// applyPolicyMapChanges applies any incremental policy map changes
+// collected on the desired policy.
+func (e *Endpoint) applyPolicyMapChanges() error {
+	errors := []error{}
+
+	//  Note that after successful endpoint regeneration the
+	//  desired and realized policies are the same pointer. During
+	//  the bpf regeneration possible incremental updates are
+	//  collected on the newly computed desired policy, which is
+	//  not fully realized yet. This is why we get the map changes
+	//  from the desired policy here.
+	adds, deletes := e.desiredPolicy.PolicyMapChanges.GetMapChanges()
+
+	for keyToAdd, entry := range adds {
+		// Keep the existing proxy port, if any
+		entry.ProxyPort = e.realizedRedirects[e.proxyID(keyToAdd)]
+		err := e.addPolicyKey(keyToAdd, entry, true)
+		if err != nil {
+			errors = append(errors, err)
 		}
 	}
 
-	err := e.addPolicyMapDelta()
+	for keyToDelete := range deletes {
+		err := e.deletePolicyKey(keyToDelete, true)
+		if err != nil {
+			errors = append(errors, err)
+		}
+	}
 
 	if len(errors) > 0 {
-		return fmt.Errorf("deleting stale PolicyMap state failed: %s", errors)
+		return fmt.Errorf("updating desired PolicyMap state failed: %s", errors)
+	} else if len(adds)+len(deletes) > 0 {
+		e.getLogger().WithFields(logrus.Fields{
+			logfields.AddedPolicyID:   adds,
+			logfields.DeletedPolicyID: deletes,
+		}).Debug("Applied policy map updates due identity changes")
 	}
 
-	return err
+	return nil
+}
+
+// syncPolicyMapDelta updates the bpf policy map state based on the
+// difference between the realized and desired policy state without
+// dumping the bpf policy map.
+func (e *Endpoint) syncPolicyMapDelta() error {
+	// Nothing to do if the desired policy is already fully realized.
+	if e.realizedPolicy != e.desiredPolicy {
+		errors := []error{}
+
+		// Delete policy keys present in the realized state, but not present in the desired state
+		for keyToDelete := range e.realizedPolicy.PolicyMapState {
+			// If key that is in realized state is not in desired state, just remove it.
+			if _, ok := e.desiredPolicy.PolicyMapState[keyToDelete]; !ok {
+				err := e.deletePolicyKey(keyToDelete, false)
+				if err != nil {
+					errors = append(errors, err)
+				}
+			}
+		}
+
+		err := e.addPolicyMapDelta()
```

Comment:
I'm a bit confused, we now have `addPolicyMapDelta()`, `syncPolicyMapDelta()` and `applyPolicyMapChanges()`.. surely we don't need three such similarly-named functions? I think there's some distinctions here between the roles of each of these, but the naming is really not helping.

Reply:
`sync` consists of adding and deleting, so `syncPolicyMapDelta()` calls `addPolicyMapDelta()`. These generate the `delta` by comparing maps. A sync without the "Delta" dumps the bpf map for the basis of comparison.

`applyPolicyMapChanges()` applies changes/deltas without the need for generating the delta by exhaustively comparing the maps.

All these functions have a different function, so we need them. trying to come up with better naming, but drawing a blank so far.

Reply:
Renaming `syncPolicyMap()` as `syncPolicyMapWithDump()` and `syncPolicyMapDelta()` as `syncPolicyMap()`. This gives a clearer indication which of them carries the cost of a full bpf map dump, and avoids using the `Delta` suffix that has a connotation with the incremental policy computation feature, while the function does a full diff between two maps.